packages/hardhat-core/test/internal/artifacts.ts

```diff
@@ -1250,6 +1261,338 @@ describe("Artifacts class", function () {
       });
     });
   });
+
+  describe("configured with custom sources", function () {
+    useTmpDir("artifacts");
+
+    it("should write and read (async) the right artifacts", async function () {
+      for (const [name, output] of Object.entries(COMPILER_OUTPUTS)) {
+        const originalArtifact = getArtifactFromContractOutput(
+          "source.sol",
+          name,
+          output
+        );
+
+        const regularArtifactSource = new Artifacts(this.tmpDir);
+        await regularArtifactSource.saveArtifactAndDebugFile(
+          originalArtifact,
+          ""
+        );
+
+        for (const CustomSource of [
+          ReadOnlySource,
+          MutableSource,
+          CachingSource,
+          Artifacts,
+        ]) {
+          const customArtifactSource = new Artifacts(
+            "bad artifact path string; internal source will fallback to custom",
+            [new CustomSource(this.tmpDir)]
+          );
```

Comment:
When `CustomSource` is `Artifacts`, this is doing something like:

```js
new Artifacts(
  "...",
  new Artifacts("this.tmpDir")
)
```

I _kind_ of understand why this works, although it surprised me. And if you look at the `Artifacts`'s constructor signature it feels like something that works more by coincidence than by design